src/lib/common/sol-modules.c

```diff
@@ -0,0 +1,222 @@
+{
+    struct module_cache *mod;
+    uint16_t i;
+
+    SOL_VECTOR_FOREACH_IDX (cache, mod, i) {
+        dlclose(mod->handle);
+        free(mod->name);
+    }
+    sol_vector_clear(cache);
+}
+
+static void
+clear_namespace_cache(struct sol_vector *cache)
+{
+    struct nspace_cache *ns;
+    uint16_t i;
+
+    SOL_VECTOR_FOREACH_IDX (cache, ns, i) {
+        clear_modules_cache(&ns->modules);
+        free(ns->name);
+    }
+    sol_vector_clear(cache);
+}
+
+static int
+get_internal_symbol(const char *symbol_name, void **symbol)
+{
+    *symbol = dlsym(RTLD_DEFAULT, symbol_name);
+    if (!*symbol)
+        return -ENOENT;
+
+    SOL_INF("Symbol '%s' found built-in", symbol_name);
+
+    return 0;
+}
+
+static struct nspace_cache *
+get_namespace(const char *nspace)
+{
+    struct nspace_cache *ns;
+    uint16_t i;
+
+    SOL_VECTOR_FOREACH_IDX (&namespaces, ns, i) {
+        if (streq(ns->name, nspace))
+            return ns;
+    }
+
+    ns = sol_vector_append(&namespaces);
+    SOL_NULL_CHECK(ns, NULL);
+
+    ns->name = strdup(nspace);
+    SOL_NULL_CHECK_GOTO(ns->name, strdup_error);
+
+    sol_vector_init(&ns->modules, sizeof(struct module_cache));
+
+    return ns;
+
+strdup_error:
+    free(ns);
+    return NULL;
+}
+
+static void *
+get_module_handle(const char *nspace, const char *modname)
+{
+    static char rootdir[PATH_MAX] = { };
+    char path[PATH_MAX];
+    int ret;
+
+    if (unlikely(!*rootdir)) {
+        ret = sol_util_get_rootdir(rootdir, sizeof(rootdir));
+        SOL_INT_CHECK(ret, >= (int)sizeof(rootdir), NULL);
+    }
+
+    ret = snprintf(path, sizeof(path), "%s%s%s/%s.so", rootdir, MODULESDIR, nspace, modname);
+    SOL_INT_CHECK(ret, >= (int)sizeof(path), NULL);
+
+    SOL_INF("Loading module '%s'", path);
+
+    return dlopen(path, RTLD_LAZY | RTLD_LOCAL | RTLD_NODELETE);
+}
+
+static struct module_cache *
+get_module(struct nspace_cache *ns, const char *module)
+{
+    struct module_cache *mod;
+    uint16_t i;
+
+    SOL_VECTOR_FOREACH_IDX (&ns->modules, mod, i) {
+        if (streq(mod->name, module)) {
+            SOL_INF("Module '%s/%s' found cached", ns->name, module);
+            return mod;
+        }
+    }
+
+    mod = sol_vector_append(&ns->modules);
+    SOL_NULL_CHECK(mod, NULL);
+
+    mod->name = strdup(module);
+    SOL_NULL_CHECK_GOTO(mod->name, strdup_error);
+
+    mod->handle = get_module_handle(ns->name, mod->name);
+    SOL_NULL_CHECK_GOTO(mod->handle, dlopen_error);
+
+    return mod;
+
+dlopen_error:
+    free(mod->name);
+strdup_error:
+    free(mod);
+    return NULL;
+}
+
+static int
+get_module_symbol(const char *nspace, const char *modname, const char *symbol_name, void **symbol)
+{
+    struct nspace_cache *ns;
+    struct module_cache *mod;
+
+    ns = get_namespace(nspace);
+    SOL_NULL_CHECK(ns, -ENOMEM);
+
+    mod = get_module(ns, modname);
+    SOL_NULL_CHECK(mod, -ENOMEM);
+
+    *symbol = dlsym(mod->handle, symbol_name);
+    SOL_NULL_CHECK(*symbol, -ENOENT);
```

Comment:
`SOL_WRN + path + symbol_name + dlerror()`. You can recreate `path` from `nspace` and `modname`